File: js/src/materialize.ts
// Reads the flat buffer `colbin::materialize::to_buffer` writes and turns it
// into row objects without going through JSON text — PACKAGE_PLAN.md §5.
//
// The buffer has no offset table: every run's byte length is derivable from
// `rowCount` and the field's own kind, so this walks it once, in field order,
// the same way the Rust writer laid it out. See `rust/src/materialize.rs`'s
// module doc for the exact layout.

const KIND_BOOL = 0
const KIND_INT = 1
const KIND_FLOAT = 2
const KIND_STRING = 3

const FLAG_SIGNED = 1 << 0
const FLAG_EXCEEDS_SAFE_INT = 1 << 1
const FLAG_ASCII = 1 << 2

/**
 * What a column's values are read as. Derived purely from a field's `kind`
 * and `flags` — never from the data — which is what lets the generated row
 * builder below be cached by field signature and reused across messages from
 * the same endpoint.
 */
type ColumnKind =
  | 'bool'
  | 'intSafe'
  | 'intBigSigned'
  | 'intBigUnsigned'
  | 'float'
  | 'stringAscii'
  | 'stringUtf8'

function classify(kind: number, flags: number): ColumnKind {
  if (kind === KIND_BOOL) return 'bool'
  if (kind === KIND_INT) {
    if ((flags & FLAG_EXCEEDS_SAFE_INT) === 0) return 'intSafe'
    return (flags & FLAG_SIGNED) !== 0 ? 'intBigSigned' : 'intBigUnsigned'
  }
  if (kind === KIND_FLOAT) return 'float'
  return (flags & FLAG_ASCII) !== 0 ? 'stringAscii' : 'stringUtf8'
}

type Field = { name: string; col: ColumnKind }

/** A uniform per-row accessor, for the CSP fallback path only — the generated
 * builder never calls this, since a function call per cell is exactly the
 * cost §2.1 measured a monomorphic object literal beating by 10x. */
type Generic = { get(row: number): unknown }

/**
 * One field's data, as views over the buffer's bytes. Every reader below —
 * the generated row builder's arguments, the fallback's accessors, and the
 * public column table — is derived from these, so the cursor arithmetic that
 * walks the runs is written once.
 */
type Run =
  | { col: 'bool'; bitmap: Uint8Array }
  | { col: 'intSafe'; hi: Int32Array; lo: Uint32Array }
  | { col: 'intBigSigned'; view: BigInt64Array }
  | { col: 'intBigUnsigned'; view: BigUint64Array }
  | { col: 'float'; view: Float64Array }
  | { col: 'stringAscii'; text: string; offsets: Uint32Array }
  | { col: 'stringUtf8'; blob: Uint8Array; offsets: Uint32Array }

function pad8(at: number): number {
  return Math.ceil(at / 8) * 8
}

/** Reads every field's run at the current cursor, in header order. */
function readRuns(buffer: ArrayBuffer, start: number, rowCount: number, fields: Field[]): Run[] {
  const runs: Run[] = []
  let at = start

  for (const field of fields) {
    switch (field.col) {
      case 'bool': {
        const bytes = Math.ceil(rowCount / 8)
        runs.push({ col: 'bool', bitmap: new Uint8Array(buffer, at, bytes) })
        at += bytes
        break
      }
      case 'intSafe': {
        // Low/high Int32 pairs over the same bytes an i64 run holds. `hi` is
        // read signed and `lo` unsigned: that reconstructs any 64-bit two's
        // complement pattern exactly, whichever the field's own signedness,
        // as long as the true value's magnitude is under 2^53 — which the
        // exceeds flag guarantees for this column kind. §2.3.
        runs.push({
          col: 'intSafe',
          hi: new Int32Array(buffer, at, rowCount * 2),
          lo: new Uint32Array(buffer, at, rowCount * 2),
        })
        at += rowCount * 8
        break
      }
      case 'intBigSigned': {
        runs.push({ col: 'intBigSigned', view: new BigInt64Array(buffer, at, rowCount) })
        at += rowCount * 8
        break
      }
      case 'intBigUnsigned': {
        runs.push({ col: 'intBigUnsigned', view: new BigUint64Array(buffer, at, rowCount) })
        at += rowCount * 8
        break
      }
      case 'float': {
        runs.push({ col: 'float', view: new Float64Array(buffer, at, rowCount) })
        at += rowCount * 8
        break
      }
      case 'stringAscii': {
        const offsets = new Uint32Array(buffer, at, rowCount + 1)
        const blobStart = at + (rowCount + 1) * 4
        const blobLen = offsets[rowCount]
        // One bulk decode over the whole blob, then `substring` per row: 10-12x
        // faster than decoding each string, because byte offsets equal UTF-16
        // code unit offsets exactly when the blob is ASCII. §5.3.
        const text = new TextDecoder().decode(new Uint8Array(buffer, blobStart, blobLen))
        runs.push({ col: 'stringAscii', text, offsets })
        at = blobStart + blobLen
        break
      }
      case 'stringUtf8': {
        const offsets = new Uint32Array(buffer, at, rowCount + 1)
        const blobStart = at + (rowCount + 1) * 4
        const blobLen = offsets[rowCount]
        runs.push({
          col: 'stringUtf8',
          blob: new Uint8Array(buffer, blobStart, blobLen),
          offsets,
        })
        at = blobStart + blobLen
        break
      }
    }
    at = pad8(at)
  }
  return runs
}

/** The flattened argument list the generated builder wants: one or two typed
 * arrays per field, in field order, matching `getFactory`'s parameter list. */
function argsOf(runs: Run[]): unknown[] {
  const args: unknown[] = []
  for (const run of runs) {
    switch (run.col) {
      case 'bool':
        args.push(run.bitmap)
        break
      case 'intSafe':
        args.push(run.hi, run.lo)
        break
      case 'intBigSigned':
      case 'intBigUnsigned':
      case 'float':
        args.push(run.view)
        break
      case 'stringAscii':
        args.push(run.text, run.offsets)
        break
      case 'stringUtf8':
        args.push(run.blob, run.offsets)
        break
    }
  }
  return args
}

/** A uniform reader per field, for the CSP fallback. */
function genericOf(runs: Run[]): Generic[] {
  return runs.map((run): Generic => {
    switch (run.col) {
      case 'bool':
        return { get: (i) => ((run.bitmap[i >> 3] >> (i & 7)) & 1) === 1 }
      case 'intSafe':
        return { get: (i) => run.hi[2 * i + 1] * 4294967296 + run.lo[2 * i] }
      case 'intBigSigned':
      case 'intBigUnsigned':
      case 'float':
        return { get: (i) => run.view[i] }
      case 'stringAscii':
        return { get: (i) => run.text.substring(run.offsets[i], run.offsets[i + 1]) }
      case 'stringUtf8': {
        const decoder = new TextDecoder()
        return { get: (i) => decoder.decode(run.blob.subarray(run.offsets[i], run.offsets[i + 1])) }
      }
    }
  })
}

type Factory = (rowCount: number, ...args: unknown[]) => unknown[]

const factoryCache = new Map<string, Factory>()

function signatureOf(fields: Field[]): string {
  return fields.map((f) => `${f.col}:${f.name}`).join('|')
}

/**
 * A generated, monomorphic row builder for this exact field signature,
 * compiled once and cached — §2.1 measures this at 10x a generic `for key in
 * schema` loop, because V8 gives every row the same hidden class.
 */
function getFactory(fields: Field[]): Factory {
  const key = signatureOf(fields)
  const cached = factoryCache.get(key)
  if (cached) return cached

  const params: string[] = []
  const entries: string[] = []
  fields.forEach((field, k) => {
    const prop = JSON.stringify(field.name)
    switch (field.col) {
      case 'bool': {
        const p = `b${k}`
        params.push(p)
        entries.push(`${prop}:((${p}[i>>3]>>(i&7))&1)===1`)
        break
      }
      case 'intSafe': {
        const hi = `h${k}`
        const lo = `l${k}`
        params.push(hi, lo)
        entries.push(`${prop}:${hi}[2*i+1]*4294967296+${lo}[2*i]`)
        break
      }
      case 'intBigSigned':
      case 'intBigUnsigned': {
        const p = `v${k}`
        params.push(p)
        entries.push(`${prop}:${p}[i]`)
        break
      }
      case 'float': {
        const p = `f${k}`
        params.push(p)
        entries.push(`${prop}:${p}[i]`)
        break
      }
      case 'stringAscii': {
        const s = `s${k}`
        const o = `o${k}`
        params.push(s, o)
        entries.push(`${prop}:${s}.substring(${o}[i],${o}[i+1])`)
        break
      }
      case 'stringUtf8': {
        const s = `s${k}`
        const o = `o${k}`
        params.push(s, o)
        entries.push(`${prop}:__dec.decode(${s}.subarray(${o}[i],${o}[i+1]))`)
        break
      }
    }
  })

  const body =
    `const __dec=new TextDecoder();` +
    `return function(n,${params.join(',')}){` +
    `const out=new Array(n);` +
    `for(let i=0;i<n;i++)out[i]={${entries.join(',')}};` +
    `return out};`
  // A generated function, not a string eval'd in scope: the only thing built
  // from field data is a property name, quoted through `JSON.stringify`, so
  // there is no injection surface here even though the source is assembled by
  // hand.
  // eslint-disable-next-line no-new-func, @typescript-eslint/no-implied-eval
  const factory = new Function(body)() as Factory
  factoryCache.set(key, factory)
  return factory
}

function buildGeneric(fields: Field[], generic: Generic[], rowCount: number): unknown[] {
  const out = new Array(rowCount)
  for (let i = 0; i < rowCount; i++) {
    const row: Record<string, unknown> = {}
    for (let k = 0; k < fields.length; k++) {
      row[fields[k].name] = generic[k].get(i)
    }
    out[i] = row
  }
  return out
}

/** Undefined until the first real probe; `allowGenerated` bypasses the cache
 * entirely, which is what lets a test force the CSP fallback without one. */
let generatedAllowed: boolean | undefined

function canUseGenerated(force?: boolean): boolean {
  if (force !== undefined) return force
  if (generatedAllowed === undefined) {
    try {
      // eslint-disable-next-line no-new-func, @typescript-eslint/no-implied-eval
      new Function('return 1')()
      generatedAllowed = true
    } catch {
      generatedAllowed = false
    }
  }
  return generatedAllowed
}

export type ReadRowsOptions = {
  /** Forces the CSP fallback path (or forces it off) regardless of what this
   * page's own CSP allows — for testing the fallback without one. */
  allowGenerated?: boolean
}

/** The buffer's header: how many rows, what the fields are, and where the
 * first run starts. */
function readHeader(buffer: ArrayBuffer): { rowCount: number; fields: Field[]; at: number } {
  const view = new DataView(buffer)
  const version = view.getUint32(0, true)
  if (version !== 1) {
    throw new Error(`materialize: unknown buffer version ${version}`)
  }
  const rowCount = view.getUint32(4, true)
  const fieldCount = view.getUint32(8, true)

  const nameDecoder = new TextDecoder()
  let at = 12
  const fields: Field[] = []
  for (let i = 0; i < fieldCount; i++) {
    const kind = view.getUint8(at)
    const flags = view.getUint8(at + 1)
    const nameLen = view.getUint16(at + 2, true)
    const name = nameDecoder.decode(new Uint8Array(buffer, at + 4, nameLen))
    at += 4 + nameLen
    fields.push({ name, col: classify(kind, flags) })
  }
  return { rowCount, fields, at: pad8(at) }
}

/**
 * The materialize buffer to row objects.
 *
 * `buffer` should be a copy the caller owns outright (not a live view into
 * wasm memory) — this constructs several typed-array views over it at
 * differing offsets, and a later `alloc()` growing the module's memory would
 * detach every one of them.
 */
export function readRows(buffer: ArrayBuffer, options: ReadRowsOptions = {}): unknown[] {
  const { rowCount, fields, at } = readHeader(buffer)
  const runs = readRuns(buffer, at, rowCount, fields)

  if (canUseGenerated(options.allowGenerated)) {
    try {
      return getFactory(fields)(rowCount, ...argsOf(runs))
    } catch {
      // A signature this engine refused to compile mid-session (a CSP that
      // only some navigations enforce, say) — fall through to the generic path.
    }
  }
  return buildGeneric(fields, genericOf(runs), rowCount)
}

/** What a column's values arrive as. */
export type Column =
  /** One value per row, unpacked from the bitmap the buffer carries. */
  | { name: string; kind: 'bool'; length: number; values: boolean[] }
  /** Every value fits a `Number` exactly — the column's own range says so. */
  | { name: string; kind: 'int'; length: number; values: Float64Array }
  /** Some value is past 2^53, so the column is carried whole rather than
   * rounded. A view over the buffer's own bytes. */
  | { name: string; kind: 'bigint'; length: number; values: BigInt64Array | BigUint64Array }
  /** A view over the buffer's own bytes. */
  | { name: string; kind: 'float'; length: number; values: Float64Array }
  | { name: string; kind: 'string'; length: number; values: string[] }

export type ColumnTable = { rowCount: number; columns: Column[] }

/**
 * The materialize buffer as columns rather than rows — PACKAGE_PLAN.md §6, for
 * a caller feeding a grid or a chart that never wants row objects.
 *
 * `float` and `bigint` columns are views straight over `buffer`; the rest are
 * built, because a bitmap, a UTF-8 blob and a pair of i32 lanes are not things
 * a chart can read. Building them is the cheap end of §2.3 — thousandths of a
 * millisecond per thousand rows — and it is still less work than the row build
 * this exists to skip.
 */
export function readColumns(buffer: ArrayBuffer): ColumnTable {
  const { rowCount, fields, at } = readHeader(buffer)
  const runs = readRuns(buffer, at, rowCount, fields)

  const columns = runs.map((run, k): Column => {
    const name = fields[k].name
    switch (run.col) {
      case 'bool': {
        const values = new Array<boolean>(rowCount)
        for (let i = 0; i < rowCount; i++) values[i] = ((run.bitmap[i >> 3] >> (i & 7)) & 1) === 1
        return { name, kind: 'bool', length: rowCount, values }
      }
      case 'intSafe': {
        const values = new Float64Array(rowCount)
        for (let i = 0; i < rowCount; i++) values[i] = run.hi[2 * i + 1] * 4294967296 + run.lo[2 * i]
        return { name, kind: 'int', length: rowCount, values }
      }
      case 'intBigSigned':
      case 'intBigUnsigned':
        return { name, kind: 'bigint', length: rowCount, values: run.view }
      case 'float':
        return { name, kind: 'float', length: rowCount, values: run.view }
      case 'stringAscii':
      case 'stringUtf8': {
        const read = genericOf([run])[0]
        const values = new Array<string>(rowCount)
        for (let i = 0; i < rowCount; i++) values[i] = read.get(i) as string
        return { name, kind: 'string', length: rowCount, values }
      }
    }
  })

  return { rowCount, columns }
}

File: rust/src/materialize.rs
//! A second sink beside [`crate::json`]: columns into a flat buffer instead of
//! decimal text, for a caller that can build its own objects from typed views.
//!
//! Mirrors `PACKAGE_PLAN.md` §5 in the browser package this crate feeds. Only
//! the table shape is covered — a root wrapped in the one-field envelope
//! (`Plan::is_envelope`), whose field is an array of structs the wire actually
//! encoded as a transposed table ([`crate::wire::Reader::is_table`]). Every
//! other shape returns `Ok(None)`, which tells the caller to fall back to
//! [`crate::walk::to_json`]: a table row can only hold a columnable op — see
//! [`plan::columnable_op`] — so a table never needs a null bitmap, a nested
//! struct or a map here, and the general document walk stays the one place
//! that handles those.
//!
//! # The buffer has no offset table
//!
//! Every run's byte length is derivable from `rowCount` and the field's own
//! kind — a bitmap is `ceil(rows/8)` bytes, an int or float run is `rows*8`,
//! and a string run's blob length is its own offset table's last entry. So the
//! reader walks the buffer once, in field order, computing each run's length
//! as it goes, rather than trusting a second copy of lengths that could
//! disagree with the data. See the module's `write_buffer` for the exact
//! layout, which the JavaScript reader mirrors.

use alloc::vec::Vec;

use crate::Error;
use crate::plan::{self, OP_STRING, OP_STRUCTS, Plan};
use crate::section::Schema;
use crate::walk::{Gathered, MAX_ROWS};
use crate::wire::{Reader, Reader8};

/// The largest integer magnitude a JavaScript `Number` holds exactly. Above
/// this the reader must pay for a `BigInt`, and the flag says so per column
/// rather than the reader guessing from the value itself.
const MAX_SAFE_INT: u64 = 9_007_199_254_740_991;

const KIND_BOOL: u8 = 0;
const KIND_INT: u8 = 1;
const KIND_FLOAT: u8 = 2;
const KIND_STRING: u8 = 3;

const FLAG_SIGNED: u8 = 1 << 0;
const FLAG_EXCEEDS_SAFE_INT: u8 = 1 << 1;
const FLAG_ASCII: u8 = 1 << 2;

/// One field's data, already in the shape its wire run will take.
enum Column {
    Bool(Vec<u8>),
    Int {
        bytes: Vec<u8>,
        signed: bool,
        exceeds: bool,
    },
    Float(Vec<u8>),
    String {
        offsets: Vec<u8>,
        blob: Vec<u8>,
        ascii: bool,
    },
}

impl Column {
    fn kind_and_flags(&self) -> (u8, u8) {
        match self {
            Column::Bool(_) => (KIND_BOOL, 0),
            Column::Int {
                signed, exceeds, ..
            } => {
                let mut flags = 0;
                if *signed {
                    flags |= FLAG_SIGNED;
                }
                if *exceeds {
                    flags |= FLAG_EXCEEDS_SAFE_INT;
                }
                (KIND_INT, flags)
            }
            Column::Float(_) => (KIND_FLOAT, 0),
            Column::String { ascii, .. } => {
                let flags = if *ascii { FLAG_ASCII } else { 0 };
                (KIND_STRING, flags)
            }
        }
    }
}

/// A message's columns, written into a flat buffer — or `None` when the root
/// is not the table shape this covers, which is not an error: the caller
/// falls back to [`crate::walk::to_json`].
///
/// # Errors
///
/// Any malformed message, the same as [`crate::walk::to_json`] would refuse.
pub fn to_buffer(schema: &Schema, message: &[u8], wide: bool) -> Result<Option<Vec<u8>>, Error> {
    let plan = schema.root();
    if !plan.is_envelope || plan.fields.len() != 1 {
        return Ok(None);
    }
    let field = &plan.fields[0];
    if field.op != OP_STRUCTS {
        return Ok(None);
    }
    let at = field.sub.ok_or(Error::BadSection)?;
    let sub_plan = schema.plan(at).ok_or(Error::BadSection)?;

    let (rows, gathered) = if wide {
        let mut reader = Reader8::new(message);
        if !(reader.more() && reader.key() == field.key) {
            return Ok(None);
        }
        if !reader.is_table() {
            return Ok(None);
        }
        let (rows, mut columns) = reader.table().ok_or(Error::Truncated)?;
        if rows > MAX_ROWS {
            return Err(Error::TooManyRows);
        }
        let mut gathered = Gathered::new(sub_plan.fields.len());
        while columns.more() {
            let Some(index) = sub_plan.field_of(columns.key()) else {
                columns.skip();
                columns.err()?;
                continue;
            };
            let op = sub_plan.fields[index].op;
            gathered.take8(&mut columns, index, op, rows);
            columns.err()?;
        }
        columns.err()?;
        reader.err()?;
        (rows, gathered)
    } else {
        let mut reader = Reader::new(message);
        if !(reader.more() && reader.key() == field.key) {
            return Ok(None);
        }
        if !reader.is_table() {
            return Ok(None);
        }
        let (rows, mut columns) = reader.counted().ok_or(Error::Truncated)?;
        if rows > MAX_ROWS {
            return Err(Error::TooManyRows);
        }
        let mut gathered = Gathered::new(sub_plan.fields.len());
        while columns.more() {
            let key = columns.key();
            let index = sub_plan.field_of(key).ok_or(Error::UnknownKey(key))?;
            let op = sub_plan.fields[index].op;
            gathered.take(&mut columns, index, op, rows);
            columns.err()?;
        }
        columns.err()?;
        reader.err()?;
        (rows, gathered)
    };

    let columns = build_columns(sub_plan, &gathered, rows)?;
    Ok(Some(write_buffer(sub_plan, rows, &columns)))
}

/// One gathered column, transformed to the value it means — the same
/// reinterpretation `Walker::column_value` does for the text sink, applied to
/// every row instead of one.
fn build_columns(table: &Plan, gathered: &Gathered, rows: usize) -> Result<Vec<Column>, Error> {
    let mut columns = Vec::with_capacity(table.fields.len());
    for (index, field) in table.fields.iter().enumerate() {
        let present = gathered.present[index];
        let raw_at = |row: usize| -> i64 {
            if present {
                gathered.ints[index].get(row).copied().unwrap_or(0)
            } else {
                0
            }
        };

        let column = match field.op {
            plan::OP_BOOL => {
                let mut bitmap = alloc::vec![0u8; rows.div_ceil(8)];
                for row in 0..rows {
                    if raw_at(row) == 1 {
                        bitmap[row >> 3] |= 1 << (row & 7);
                    }
                }
                Column::Bool(bitmap)
            }
            op if plan::is_integer_op(op) => {
                let signed = plan::is_signed_op(op);
                let width = plan::width_of_op(op);
                let mut bytes = Vec::with_capacity(rows * 8);
                let mut exceeds = false;
                for row in 0..rows {
                    let raw = raw_at(row);
                    let value: i64 = if signed {
                        if raw.unsigned_abs() > MAX_SAFE_INT {
                            exceeds = true;
                        }
                        raw
                    } else {
                        // Same truncation `column_value` applies: the column
                        // codec stores signed residuals, so a column narrower
                        // than 64 bits comes back sign-extended.
                        let value = if width == 8 {
                            raw as u64
                        } else {
                            (raw as u64) & (u64::MAX >> (64 - width * 8))
                        };
                        if value > MAX_SAFE_INT {
                            exceeds = true;
                        }
                        value as i64
                    };
                    bytes.extend_from_slice(&value.to_le_bytes());
                }
                Column::Int {
                    bytes,
                    signed,
                    exceeds,
                }
            }
            plan::OP_FLOAT32 => {
                let mut bytes = Vec::with_capacity(rows * 8);
                for row in 0..rows {
                    let value = f64::from(f32::from_bits(raw_at(row) as u32));
                    if value.is_nan() || value.is_infinite() {
                        return Err(Error::NotJson);
                    }
                    bytes.extend_from_slice(&value.to_le_bytes());
                }
                Column::Float(bytes)
            }
            plan::OP_FLOAT64 => {
                let mut bytes = Vec::with_capacity(rows * 8);
                for row in 0..rows {
                    let value = f64::from_bits(raw_at(row) as u64);
                    if value.is_nan() || value.is_infinite() {
                        return Err(Error::NotJson);
                    }
                    bytes.extend_from_slice(&value.to_le_bytes());
                }
                Column::Float(bytes)
            }
            OP_STRING => {
                let mut offsets = Vec::with_capacity((rows + 1) * 4);
                let mut blob = Vec::new();
                let mut ascii = true;
                offsets.extend_from_slice(&0u32.to_le_bytes());
                for row in 0..rows {
                    let value: &[u8] = if present {
                        gathered.strings[index].get(row).copied().unwrap_or(&[])
                    } else {
                        &[]
                    };
                    if ascii && !value.is_ascii() {
                        ascii = false;
                    }
                    blob.extend_from_slice(value);
                    let end = u32::try_from(blob.len()).map_err(|_| Error::SizeTooLarge)?;
                    offsets.extend_from_slice(&end.to_le_bytes());
                }
                Column::String {
                    offsets,
                    blob,
                    ascii,
                }
            }
            // `Plan::finish` refuses a table whose element is not fully
            // columnable, so nothing else can reach here.
            op => return Err(Error::UnwalkableOp(op)),
        };
        columns.push(column);
    }
    Ok(columns)
}

fn pad_to_8(out: &mut Vec<u8>) {
    let pad = out.len().next_multiple_of(8) - out.len();
    out.resize(out.len() + pad, 0);
}

/// The layout the module doc describes: a header naming every field, then one
/// data run per field in the same order, each aligned to 8 bytes so a
/// `Float64Array`/`BigInt64Array` view over it needs no copy.
fn write_buffer(table: &Plan, rows: usize, columns: &[Column]) -> Vec<u8> {
    let mut out = Vec::new();
    out.extend_from_slice(&1u32.to_le_bytes()); // version
    out.extend_from_slice(&(rows as u32).to_le_bytes());
    out.extend_from_slice(&(table.fields.len() as u32).to_le_bytes());

    for (name, column) in table.names.iter().zip(columns) {
        let (kind, flags) = column.kind_and_flags();
        out.push(kind);
        out.push(flags);
        let name_bytes = name.as_bytes();
        out.extend_from_slice(&(name_bytes.len() as u16).to_le_bytes());
        out.extend_from_slice(name_bytes);
    }
    pad_to_8(&mut out);

    for column in columns {
        match column {
            Column::Bool(bitmap) => out.extend_from_slice(bitmap),
            Column::Int { bytes, .. } | Column::Float(bytes) => out.extend_from_slice(bytes),
            Column::String { offsets, blob, .. } => {
                out.extend_from_slice(offsets);
                out.extend_from_slice(blob);
            }
        }
        pad_to_8(&mut out);
    }
    out
}
